pydicom/dataset.py

```diff
@@ -653,6 +732,107 @@ def __getitem__(self, key):
 
         return self._dict.get(tag)
 
+    def private_block(self, group, private_creator, create=True):
+        """Return the block for the given tag and private creator.
+
+        If `create` is set and the private creator does not exist,
+        the private creator tag is added.
+        Note: We ignore the unrealistic case that no free block is
+        available.
+
+        Parameters
+        ----------
+        group : 32 bit int
+            The group of the private tag to be found. Must be an odd number
+            (e.g. a private group).
+        private_creator : str
+            The private creator string associated with the tag.
+        create : bool
+            If `True` and `private_creator` does not exist, a new private
+            creator tag is added at the next free block.
+
+        Returns
+        -------
+        32 bit int
+            Element base for the given tag (the last 2 hex digits are always 0)
+
+        Raises
+        ------
+        ValueError
+            If `tag` is not a private tag or `private_creator` is empty.
+        KeyError
+            If the private creator tag is not found in the given group and
+            the `create` parameter is not set.
+        """
+        def new_block():
+            block = PrivateBlock(key, self, element)
+            self._private_blocks[key] = block
+            return block
+
+        key = (group, private_creator)
+        if key in self._private_blocks:
+            return self._private_blocks[key]
+
+        if not private_creator:
+            raise ValueError('Private creator must have a value')
+
+        if group % 2 == 0:
+            raise ValueError(
+                'Tag must be private if private creator is given')
+
+        for element in range(0x10, 0x100):
+            private_creator_tag = Tag(group, element)
+            if private_creator_tag not in self._dict:
+                if create:
+                    self.add_new(private_creator_tag, 'LO', private_creator)
+                    return new_block()
+                else:
+                    break
+            if self._dict[private_creator_tag].value == private_creator:
+                return new_block()
+
+        raise KeyError(
+            "Private creator '{}' not found".format(private_creator))
+
+    def get_private_item(self, key, private_creator):
+        """Return the data element for the given private tag.
```

Comment:
I'm not sure about this one. I see these possibilities:
- leave as is
- change to `get_private_item(self, group, element_offset, private_creator)`
- change usage to:
```python
block = ds.private_block(group, private_creator, create=False) # may add a convenience wrapper for this
item = block.get_item(element_offset)
```
What do you think?

Reply:
I like the last one - getting the block and then accessing elements. That just fits my brain the best.  However. it could co-exist with the second one. 

But the create=False seems awkward.  What do you think about making the create step its own function? E.g.
```
ds.create_private_block(group, private_creator)
```
Maybe that makes sense now that the 'get' side has the info  cached by `dataset._private_blocks`, so there's no real code overlap.  Is this something like what you meant about a wrapper?

Reply:
I actually just meant to hide the `create` argument in another wrapper function.
I'm not sure I understand your proposal:  should `create_private_block` create a new block, so you have to do something like:
```python
  block = ds.get_private_block(group, private_creator)
  if block is None:
      block = ds.create_private_block(group, private_creator)
  block.add_new(...)
```
Probably not, as that would be too complicated... Can you please describe the usage you have in mind? I'm obviously missing something here...

Reply:
Well, that is pretty much what I was saying, but I was thinking with catching an exception.  I know it is a bit awkward ... but not out of line with the way many things are done, including regular data element access.  Could also do it with an "in" function:
```python
if not 'Creator name' in ds.private_blocks:
   ...
```
Again, similar to how data elements are accessed.

Another thought:  could we invert the create default?  Make creation the special one, and not-creating the default.  I think that was what you originally had.

Reply:
Ok, thanks! 

> Another thought: could we invert the create default? 
Yes, I had changed that because the `False` case had not been needed in a public function, which changes now, so I agree.

Though, if I leave that parameter in, I could still use the current syntax:
```python
  block = ds.private_block(group, private_creator, create=True)
  block.add_new(...)
```
and would not need the separate `create_private_block()`, which has not much value to the user IMHO, as the user will not need private blocks without tags. Having said that, there actually could be cases where a private creator shall be added as a marker, so it may still make sense...

I like the idea of the `in` check - but that would also mean that we either have to keep track of each private block while reading data, or we should just recheck for private blocks in `Dataset.private_blocks`, as they are created on demand only.

I think I'll make a few changes (add the getter in `PrivateBlock`, change the default parameter, and change the arguments for `get_private_item`, and we can go from there.

Reply:
> Though, if I leave that parameter in, I could still use the current syntax

Yes, agreed.  I think that works well.



> I like the idea of the `in` check - but that would also mean that we either have to keep track of each private block while reading data, or we should just recheck for private blocks in `Dataset.private_blocks`, as they are created on demand only

I'm not sure if we are saying the same thing, but I was thinking that `in` could just be a convenience to catching the exception from `ds.private_block()` with create as False, and turning the exception (or not) into a boolean.  I think this matches your 'on demand only' comment - the internal list wouldn't have all private blocks that might exist.

Reply:
Ah, ok. I hadn't realized that `Dataset.private_blocks` would be a new function. Should that be returning the private creators for the given group? In this case, may be we better name it `private_creators'?

Reply:
> Should that be returning the private creators for the given group?

Yes, that makes good sense.  And that's made me think of something else: we could use `ds.group_dataset(group).keys()` to get a list of all data element tags under the group, which could be filtered in a list comprehension for creators.  Perhaps those keys could also be used instead of the loop?

Reply:
Hm, not sure about this. `group_dataset` does:
`self[(group, 0x0000):(group + 1, 0x0000)]`
We actually need:
`self[(group, 0x0000):(group, 0x0100)]`
so I would use this directly instead of getting a dataset for the whole group and then filtering out all non-private-creator tags, if we need the dataset subset. I'm not sure I understand the reasoning behind this, though (I'm a bit slow today understanding anything, apparently...)

What came to my mind tht may be somewhat related, was that it would be good if we could ask the block for it's keys, to be able to write:
```python
  if 'My Creator' in ds.private_creators(group):
      block = ds.private_block(group)
      if 0x05 in block:
          my_value = block.get_item(0x05).value
```

Reply:
Also, it may be more consistent to use `__getitem__` instead of `get_item` in `PrivateBlock`, so that we would use the block similar to a Dataset:
```python
  if 0x05 in block:
      my_value = block[0x05].value
```

Reply:
Ok, I changed this (e.g. `get_item` -> `__getitem__`) - will get back to this tomorrow.

Reply:
I also added item deletion, and am basically done now (except for the change related to `ds.group_dataset(group).keys()`, which I didn't understand yet). Please check what is still missing.

Reply:
> `ds.group_dataset(group).keys()`, which I didn't understand yet)

I think that is no longer relevant -- you were of course correct that slicing directly is better.

I think these changes are really nice ... will review code in detail later today or tomorrow.